pandas/_libs/reduction.pyx

```diff
@@ -59,16 +59,20 @@ cdef class _BaseGrouper:
             cached_typ = self.typ(
                 vslider.buf, dtype=vslider.buf.dtype, index=cached_ityp, name=self.name
             )
+            self.has_block = hasattr(cached_typ._mgr, "_block")
         else:
             # See the comment in indexes/base.py about _index_data.
             # We need this for EA-backed indexes that have a reference
             # to a 1-d ndarray like datetime / timedelta / period.
             object.__setattr__(cached_ityp, '_index_data', islider.buf)
             cached_ityp._engine.clear_mapping()
             cached_ityp._cache.clear()  # e.g. inferred_freq must go
-            object.__setattr__(cached_typ._mgr._block, 'values', vslider.buf)
-            object.__setattr__(cached_typ._mgr._block, 'mgr_locs',
-                               slice(len(vslider.buf)))
+            if self.has_block:
+                object.__setattr__(cached_typ._mgr._block, 'values', vslider.buf)
+                object.__setattr__(cached_typ._mgr._block, 'mgr_locs',
+                                   slice(len(vslider.buf)))
+            else:
+                cached_typ._mgr.arrays[0] = vslider.buf
```

Comment:
I could probably replace this full if/else block with a single `cached_typ._mgr.set_values(vslider.buf)` (if I add the setting of `mgr_locs` to `SingleBlockManager.set_values`). 

But I assume that, currently, we don't use a plain python attribute setting but `object.__setattr__` for performance? Using `_mgr.set_values(..)` might defeat that purpose?

Reply:
So the newer commit I pushed does this change, and is thus now using `_mgr.set_values`. 

I did some timings with a specific function that uses the SeriesGrouper with many labels + a cheap dummy function (adapted from the same benchmark case as I have been using for the other groupby PRs: https://github.com/pandas-dev/pandas/pull/40178#issue-583301313):

```
ncols = 1000
N = 1000
data = np.random.randn(N, ncols)
labels = np.random.randint(0, 100, size=N)
df = pd.DataFrame(data)

%timeit df.groupby(labels)[0].agg(lambda x: 1)
```

And repeating this several times switching back and forth between this version and master, I don't see any difference. A representative timing was:

```
In [15]: %timeit df.groupby(labels)[0].agg(lambda x: 1)
1.18 ms ± 111 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  <-- master
1.17 ms ± 88.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  <-- PR
```